backend/nodes/image_dim_nodes.py

```diff
@@ -27,8 +27,14 @@ def __init__(self):
         )
         self.inputs = [
             ImageInput(),
-            BoundedNumberInput(
-                "Scale Factor (%)", minimum=0.1, maximum=None, default=100.0, step=25.0
+            NumberInput(
+                "Scale Factor",
+                precision=4,
+                step=0.0001,
```

Comment:
Is it even necessary to specify a step here? Couldn't we just say that the default for step is `10 ** -precision`?

Reply:
Zamboni wants the step to be different sometimes. Like having it be 0.25 even though you could make the value 0.05 if you wanted. I kinda personally think that might be a bit confusing though

Reply:
Not what I meant. Here, he made step its smallest possible value `10 ** -4`. Values smaller than this would conflict with precision. I don't mean the control step, that is a good addition IMO.

Honestly, the underlying problem here is that `precision` and `step` are very similar. They both limit the number of possible inputs within the input range, they just do it through different means. `precision` limits the number of possible inputs by allowing only a certain number of decimal places. `step` is more general and makes that sure all input values are of the form `offset + i * step, i∈Z` (ignoring minimum and maximum for now).

Any precision value can be represented by a step of `step = 10 ** -precision` with `offset = 0`, but not every step value can be represented by a precision. E.g. it is impossible to convert the step value 1/3 into a precision other than ∞.

It might be a good idea to get rid of precision completely, and just have a little helper function `precision_to_step`.

The only one that needs precision is the frontend, and precision can be computed from offset and step as `max(getPrecision(offset), getPrecision(step))`. The `getPrecision` function itself could be implemented like this:

```ts
function getPrecision(n: number): number {
	const fraction = n % 1;
	if (fraction === 0) return 0;
	if (Math.abs(fraction) < 1e-10) return 10;
	return Math.min(10, String(n).split('.')[1]?.length ?? 0);
}
```

Note that this implementation will return at most 10. 64bit floating numbers only have a precision of around 16 digits anyway, so I chose 10, because it's enough for most use cases and it hides away any floating point inaccuracies.

Reply:
I actually decided after I was done for the night that I should ditch precision and keep step, so I'm already planning to do this.

Reply:
I think it would be better to leave precision as a derived attribute and calculate it in the backend, though, since the backend validation requires the value as well.

Reply:
`value = (
    round((round((value - offset) / step) * step + offset) * 10**precision)
    * 10**-precision
)`

It is used here. Because step could be something like 0.5, you can't simply multiply the value by step, so you need to perform mathematical operations on step regardless to use it.

Reply:
Also, it's just a one liner in python:
`self.precision = abs(Decimal(str(step)).as_tuple().exponent)`

Reply:
> It is used here.

I know that `precision` is used in the current formula... Me point is that this can be removed. 

> Because step could be something like 0.5, you can't simply multiply the value by step, so you need to perform mathematical operations on step regardless to use it.

I don't get what you mean. 

Also, "mathematical operations" are literally all that computers can do. You're gonna have to be a bit more specific.

> Given that, it makes more sense to do it once in one place and simply pass the value around.

It doesn't. From my perspective, it does not make sense. As I've shown above, `step`+`offset` is strictly more powerful than `precision`. 

You can literally convert all `precision`,`step`,`offset` triples into equivalent `step`,`offset` pairs. So why would we even bother with `precision`?

Reply:
> Also, "mathematical operations" are literally all that computers can do. You're gonna have to be a bit more specific.

The mathematical operations that have to be performed to transform step into a precision value, e.g. that getPrecision function that you wrote above. This needs to be done in both the frontend *and* the backend, as the backend performs its own value correction in case some invalid value slips through the frontend. If we only use step or step+offset, then the getPrecision function will need to be performed by the frontend, then again by an equivalent function in the backend. On the other hand, if we just include a precision attribute and calculate it from step or step+offset at object initialization in the backend, this only has to be calculated one time.

Reply:
> as the backend performs its own value correction in case some invalid value slips through the frontend

As I said before, `precision` is not necessary on the backend. All these corrections can be done without `precision`.

Reply:
`precision` is literally only needed in the frontend for the numeric input component. 

A `step`-`offset` pair is enough for the correction, again, as I said before. To be more specific, the formula is this:

```
corrected_value = round((value - offset) / step) * step + offset
```

`precision` is just unnecessary complexity.

Reply:
> As I said before, precision is not necessary on the backend. All these corrections can be done without precision.

Then what is that equation? There may be one, so if you know it, just write it out instead of repeating that it can be done.